design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,414 @@
+field. However, there is an increasing demand to store infrastructure
+credentials on external secret stores rather than relying on Kubernetes secrets.
+
+Main motivations behind this demand are:
+- Security: Kubernetes secrets are known to be insecure since they store data
+as an unencrypted base64-encoded string. Proper access control via RBAC is
+usually too hard if not impossible.
+- Reliability: Most users consider Kubernetes clusters as ephemeral resources
+which could be replaced with a new one (e.g. via GitOps) for some reason.
+Using on Kubernetes API to store infrastructure credentials is not a reliable
+approach in this sense.
+
+In this document we will investigate alternative options to support additional
+secret stores in Crossplane. Throughout the document, we will mostly focus on
+[Vault] as the most popular secret store today.
+
+## Goals
+
+The goal of this document is to investigate alternative approaches to support
+external secret stores in Crossplane to store connection details. The solution
+should apply to both Crossplane itself (for composite connection details) and
+providers (for managed resource connection details).
+
+## Out of Scope
+
+This document focuses on writing/publishing connection details to external
+stores and not about reading/consuming them in the form of provider credentials
+or to consume via the workloads. There are already ways to consume secrets in
+external stores from Kubernetes and is out of the scope for this document.
+
+See [Vault credential injection guide] to see how one can configure Vault and
+Crossplane to consume provider credentials from Vault. To consume secrets living
+inside Vault from workloads running in Kubernetes, one can consider
+[injecting via Vault agent] or use solutions like [Vault secrets operator].
+
+## Design
+
+There are a couple of alternative approaches that we would affect the design.
+We would like to list them all here and finally agree on one of them.
+
+### Option 1: Supporting Additional Stores in Crossplane-runtime
+
+This is the most straightforward approach. We will implement support for writing
+to external stores in _crossplane-runtime_ via importing and using their 
+clients. With this solution, Crossplane and providers would immediately get
+this feature when configured properly.
+
+![Secret Store Runtime][secret-store-runtime]
+
+#### API
+
+We will deprecate existing `writeConnectionSecretToRef` field in favor of a
+`publishConnectionDetailsTo` field which would support configuring multiple
+secret backends. This will apply to not only managed resources but also to
+_Composite Resources_ and _Composite Resource Claims_.
+
+```
+apiVersion: widgets.crossplane.io/v1beta1
+kind: Example
+metadata:
+  name: my-cool-managed-resource
+spec:
+  forProvider:
+    coolness: 24
+  publishConnectionDetailsTo:
+  - store: KubernetesSecret
+    kubernetesSecret:
+      secretRef:
+        name: my-cool-managed-resource-connection
+        namespace: secrets
```

Comment:
```suggestion
  - kubernetesSecretRef:
        name: my-cool-managed-resource-connection
        namespace: secrets
```

It seems like we could drop the `store` field and opt, instead, to detect the type from the name of the field submitted. I've seen this pattern already in use with our Refs and Selectors in Crossplane.

Reply:
This comment can be ignored in light of the discussion lower down about leaving Store Type determination to the StoreConfig.